### ISSUE TITLE
Download link not appearing on Your Past Conversations chat. 

### ISSUE THREAD

**Issue by Remmet577** [0x000011b/characterai-dumper#10]
The download link is not being injected into the Chat Histories page.  It worked a few days ago. It still appears on the Character Definition page. 

I am on Opera GX 96.0.4693.127 using tampermonkey.  

In the console, I can see lines like "[CharacterAI Dumper v1.4] Got character info for..." indicating the script is running.  However, when I pick through the page source, there is no element for the Download link (assuming it's supposed to exist in the title area along with "Your Past Conversations with...").

**Comment by BowNArrowPro20** [comment]
Yeah I've got the same issue on chrome.

**Comment by DreamerMemes** [comment]
I'm also having this same exact issue, perhaps the script needs an update?

**Comment by DreamerMemes** [comment]
> I'm also having this same exact issue, perhaps the script needs an update?

Yeah it has started working again for some reason, probably was something on chat ai's side.

**Comment by DreamerMemes** [comment]
> > I'm also having this same exact issue, perhaps the script needs an update?
> 
> Yeah it has started working again for some reason, probably was something on chat ai's side.

Nvm, the issue still persists, but it seems to be with the more popular bots? I'm not entirely sure. When I picked a less popular bot and checked the saved chats the download link quickly appeared, but when I got to 2B for example the issue appears. I'm going to leave it on over night and see what happens.

**Comment by Remmet577** [comment]
I'm still getting it for my private bot so it's not only the popular ones. Strange. 

(EDIT: when I say getting it I mean I'm still getting the issue i.e. no download links show up)

**Comment by DreamerMemes** [comment]
Yeah I thought it was some sort of time thing, like it just took the download button ages to appear but I'm not really sure what it is. At this point I'm pretty much confused on what it could possibly be. 

> I'm still getting it for my private bot so it's not only the popular ones. Strange.
> 
> (EDIT: when I say getting it I mean I'm still getting the issue i.e. no download links show up)

When you mentioned private bots I checked mine as well and I can't see the download link in saved chats either which has confused me even more so than before. It might be the script needs an update because something might've change on character ai's side? Of course I have no idea what I'm talking about and just making educated guesses at this point. We'll know what the problem could be if/when the creator looks at the problem themself. Here's to hoping I guess.

**Comment by gitscrub** [comment]
The site no longer sends the entire chat history. The log has been split into pages. That's why the total message count is gone as well.

The saved chats page sends you the latest two messages, and when you flip back to the full chat, you're sent the latest sixteen messages. Scrolling up will send you prior pages of _ten_ messages each, one page at a time. Unless there's still a way to access the entire history in full, if it still exists in that form, this script in its current form is dead. Without that, the only other option is to make a Discord-style script that automatically scrolls up and appends each page to a file, which would still be difficult as the messages that make up each batch of ten would change based on the total message count.

**Comment by Remmet577** [comment]
Bummer. Could you maybe change the script so that it adds a download link to each chat (from the Chat page)? You can still access the entire chat from there.  (Although I have noticed that there is a pause as you scroll up as it fetches the previous block, so maybe that's kept server side and you're unable to get it.)

**Comment by DreamerMemes** [comment]
Damn, I guess the most I'll have to use it for now is just pulling definitions and such from the bots. I guess I can deal with loosing the messages. It would be nice if it were possible to change the script so it would be able to function the same way it did before.

**Comment by 0x000011b** [comment]
Yeah, as said above it looks like CAI finally started trying to be more efficient with their serialization and now they only send small chunks of chat history, so simply intercepting the HTTP request to the histories endpoint doesn't work as expected anymore.

Unfortunately this userscript is not really a priority compared to other things I'm working on at the moment so I won't be investigating further for the time being, but I'm open to reviewing any PRs addressing this.

**Comment by Akjosch** [comment]
Couldn't you grab whatever HTML is visible? I just tried this tiny bit of code on the console (after scrolling up so the whole conversation is loaded), and that seems to work. At least for the length of the conversation I have.

```javascript
var c = document.getElementsByClassName("infinite-scroll-component")[0];
var dump = Array.from(c.getElementsByClassName("msg"))
	.flatMap((d) => Array.from(d.getElementsByTagName("p"))
		.reverse()
		.map((p) => "<p><strong>" + (d.className === "msg char-msg" ? "{char}" : "{user}") + "</strong>" + p.innerHTML + "</p>"))
	.reverse()
	.join("\n");
console.log(dump);
```

**Comment by LazerCuber** [comment]
probably gotta find ways to tweak it through console I guess...

**Comment by Loverofafoxgirl** [comment]
Will this ever be updated?

**Comment by 0x000011b** [comment]
> Will this ever be updated?

Likely not. My time is taken up by other projects at the moment, and apparently CAI Tools has already fixed their extension to allow download of CAI chats just like this userscript - so I think using that makes more sense than trying to maintain this.

Will update the README to note the project's current status to avoid confusion.
